nav2_dwb_controller/dwb_critics/test/prefer_forward_test.cpp

```diff
@@ -0,0 +1,154 @@
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <math.h>
+#include <vector>
+
+#include <memory>
+#include <string>
+#include <limits>
+
+#include "gtest/gtest.h"
+#include "rclcpp/rclcpp.hpp"
+#include "dwb_critics/prefer_forward.hpp"
+#include "nav2_costmap_2d/costmap_2d.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+
+TEST(PreferForward, StartNode)
+{
+  auto critic = std::make_shared<dwb_critics::PreferForwardCritic>();
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  auto node = nav2_util::LifecycleNode::make_shared("costmap_tester");
+  node->configure();
+  node->activate();
+  // rclcpp::spin_some(node->get_node_base_interface());
+
+  std::string name = "test";
+  std::string ns = "ns";
+  critic->initialize(node, name, ns, costmap_ros);
+  // rclcpp::spin_some(node->get_node_base_interface());
+
+  critic->onInit();
+
+  // rclcpp::spin_some(node->get_node_base_interface());
+
+  EXPECT_EQ(node->get_parameter(ns + "." + name + ".penalty").as_double(), 1.0);
+  EXPECT_EQ(node->get_parameter(ns + "." + name + ".strafe_x").as_double(), 0.1);
+  EXPECT_EQ(node->get_parameter(ns + "." + name + ".strafe_theta").as_double(), 0.2);
+  EXPECT_EQ(node->get_parameter(ns + "." + name + ".theta_scale").as_double(), 10.0);
+}
+
+TEST(PreferForward, NegativeVelocityX)
+{
+  auto critic = std::make_shared<dwb_critics::PreferForwardCritic>();
+  dwb_msgs::msg::Trajectory2D trajectory;
+
+  // score must be equal to the penalty (1.0) for any negative x velocity
+  trajectory.velocity.x = -1.0;
+  EXPECT_EQ(critic->scoreTrajectory(trajectory), 1.0);
```

Comment:
Replacing `1.0` with a local variable `penalty` would be good to be clear why it would be 1 (because this critic makes non-forward `penalty` cost). It would also let you run this with different settings of `penalty`.

Reply:
As well for the `strafe` test below